Keep a composite presentation alive when its item node is replaced (#7030)
ariakit/ariakit: .changeset/300-composite-replaced-item-presentation.md

---
"@ariakit/react-components": patch
"@ariakit/react": patch
---

> Fixed composite widgets failing to scroll to their active item when that item's element was replaced while the popup was still being positioned, which affects [`Combobox`](https://ariakit.com/reference/combobox), [`Select`](https://ariakit.com/reference/select) and [`Menu`](https://ariakit.com/reference/menu). This could happen when an app swapped in a fresh list as the popup opened, with the same item ids under new React keys, leaving the popup scrolled to the top instead of to the active item.

Author: This should be one sentence.

Author: <!-- ariakit-pr-review review_id=pr-7030-f0c9264d9d769b5c709a9a279275916718549bea-20260803T111845Z head_sha=f0c9264d9d769b5c709a9a279275916718549bea -->

Confirmed. The shared changeset guidance for a simple, single-paragraph entry is one short past-tense summary sentence ending with a period, with a final credit sentence only when contributor credit is required. This entry is two sentences, and `#7021` was opened by you, so no credit sentence is due here either.

Both sentences carry information worth keeping, so folding the trigger into the summary rather than dropping it:

```suggestion
Fixed [`Combobox`](https://ariakit.com/reference/combobox), [`Select`](https://ariakit.com/reference/select) and [`Menu`](https://ariakit.com/reference/menu) popups opening scrolled to the top instead of to the active item when an app replaced the item elements under the same ids while the popup was still being positioned.
```

That keeps every API mention linked, leads with the user-visible symptom rather than the internal cause, and stays past tense.

One dependency: the `Menu` mention is contingent on the real-focus gap raised on `packages/ariakit-react-components/src/composite/utils.ts:200`. `Menu` inherits `virtualFocus: false`, and I reproduced a real-focus menu whose replaced item is still not scrolled into view on this head. If that gap is closed, the sentence above is accurate as written. If it is left open, `Menu` should be dropped or qualified here, since the changeset guidance asks for every listed user-visible behavior to be verified against current code or a repro before keeping it.

All three reference URLs return HTTP `200`.

Author: Removed the file instead, in [`4cca022`](https://github.com/ariakit/ariakit/commit/4cca02284), so this thread now anchors to a path the branch no longer has.

Your suggestion is the right shape for a one-sentence entry, and I would have taken it, but checking the release state first turned up that there is nothing to describe. `presentItem` was introduced by #7009 (`f628a6f2`), and `git merge-base --is-ancestor f628a6f2 cbeea3cc` is false against the last `Publish` commit, so no released version can reach this behavior. #7009's own `300-composite-focus-presentation.md` is still pending on `main`, so the changelog for the next release already describes the presentation work this is part of. The released code also resolved its scroll target from the store on every `items` change, so it survived node replacement already.

The sibling fixes to the same unreleased code, #7029 and #7031, both landed without a changeset, which is the precedent I followed.

Happy to restore it with your wording if you would rather have the entry anyway.
